sys/kern/kern_ctf.c

```diff
@@ -225,7 +225,8 @@ link_elf_ctf_get(linker_file_t lf, linker_ctf_t *lc)
 	 * Read the CTF data into the raw buffer if compressed, or
 	 * directly into the CTF buffer otherwise.
 	 */
-	if ((error = vn_rdwr(UIO_READ, nd.ni_vp, raw == NULL ? ctftab : raw,
+	if ((error = vn_rdwr(UIO_READ, nd.ni_vp,
+	    PTR2CAP((void *)(raw == NULL ? ctftab : raw)),
```

Comment:
Hmm, shame we need the `void *` cast here.

Reply:
We shouldn't need a cast? PTR2CAP on that should give a `char * __capability` which can be implicitly cast to the desired `void * __capability`...

Reply:
Hmm, the `void *` indeed unnecessary, but due to a missing set of parentheses the extra set of parentheses was required (I'll fix the macro):
```
/home/bed22/cheri/cheribsd/sys/kern/kern_ctf.c:229:18: error: comparison between pointer and integer ('char' and 'void *') [-Werror,-Wpointer-integer-compare]
            PTR2CAP(raw == NULL ? ctftab : raw),
            ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
/home/bed22/cheri/cheribsd/sys/sys/param.h:433:30: note: expanded from macro 'PTR2CAP'
        (__cheri_tocap __typeof__((*p)) * __capability)(p);     \
                                   ~^
In file included from /home/bed22/cheri/cheribsd/sys/kern/link_elf.c:144:
/home/bed22/cheri/cheribsd/sys/kern/kern_ctf.c:229:6: error: invalid __cheri_tocap from 'caddr_t' (aka 'char *') to unrelated type 'typeof ((*raw == ((void *)0) ? ctftab : raw)) * __capability' (aka 'char ** __capability')
            PTR2CAP(raw == NULL ? ctftab : raw),
            ^
/home/bed22/cheri/cheribsd/sys/sys/param.h:433:49: note: expanded from macro 'PTR2CAP'
        (__cheri_tocap __typeof__((*p)) * __capability)(p);     \
                                                       ^
In file included from /home/bed22/cheri/cheribsd/sys/kern/link_elf.c:144:
/home/bed22/cheri/cheribsd/sys/kern/kern_ctf.c:229:6: error: passing 'void' to parameter of incompatible type 'void * __capability'
            PTR2CAP(raw == NULL ? ctftab : raw),
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bed22/cheri/cheribsd/sys/sys/param.h:429:20: note: expanded from macro 'PTR2CAP'
#define PTR2CAP(p)      ({                                      \
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bed22/cheri/cheribsd/sys/sys/vnode.h:780:67: note: passing argument to parameter 'base' here
int     vn_rdwr(enum uio_rw rw, struct vnode *vp, void * __capability base,
                                                                      ^
```